docs: add more explanations on how to integrate aligned (#582)
yetanotherco/aligned_layer: docs/guides/2_integration_aligned_into_your_application.md

Reviewer: Change name to `2_integrating_aligned_into_your_application.md`

# Integrating Aligned into your Application

Aligned can be integrated into your applications in a few simple steps to provide a way to verify ZK proofs generated inside your system.  

This example shows a sample app that generates an SP1 proof that a user knows the answers to a quiz, then submits the proof to Aligned for verification.
Finally, it includes a smart contract that verifies that a proof was verified in Aligned and mints an NFT.

You can find an example of the full flow of using Aligned on your app in the [ZKQuiz example](../../examples/zkquiz). 

## Steps

### Step 1 - Generate your ZK Proof

Generate your ZK proofs using any of the proving systems supported by Aligned.
For this example, we use the SP1 proving system. The current SP1 version used in Aligned is v1.0.8-testnet.

> You can find the example of the quiz proof [program](../../examples/zkquiz/quiz/program/src/main.rs) as well as the [script](../../examples/zkquiz/quiz/script/src/main.rs) that generates it in the [ZKQuiz example](../../examples/zkquiz) directory.

Reviewer: ```suggestion
You can find an example of the quiz proof [program](../../examples/zkquiz/quiz/program/src/main.rs) as well as the [script](../../examples/zkquiz/quiz/script/src/main.rs) that generates it in the [ZKQuiz example](../../examples/zkquiz) directory.
```

### Step 2 - Write your smart contract

To check if a proof was verified in Aligned, you need to call to the Aligned ServiceManager contract inside your smart contract. 

Also, you will need a way to check that the proven program is your own.

> The aligned cli provides a way for you to get the verification key commitment

Reviewer: ```suggestion
The aligned CLI provides a way for you to get the verification key commitment
```

        bytes memory merkleProof,
        uint256 verificationDataBatchIndex
    ) {
        // ... Your function code
        
        require(elfCommitment == provingSystemAuxDataCommitment, "ELF does not match");
        
        (bool callWasSuccessful, bytes memory proofIsIncluded) = alignedServiceManager.staticcall(
            abi.encodeWithSignature(
                "verifyBatchInclusion(bytes32,bytes32,bytes32,bytes20,bytes32,bytes,uint256)",
                proofCommitment,
                pubInputCommitment,
                provingSystemAuxDataCommitment,
                proofGeneratorAddr,
                batchMerkleRoot,
                merkleProof,
                verificationDataBatchIndex
            )
        );

        require(callWasSuccessful, "static_call failed");
        
        bool proofIsIncludedBool = abi.decode(proofIsIncluded, (bool));
        require(proofIsIncludedBool, "proof not included in batch");
        
        // Your function code ...
    }
}
```

You can find the example of the smart contract that checks the proof was verified in Aligned
in the [Quiz Verifier Contract](../../examples/zkquiz/contracts/src/VerifierContract.sol).

> Note that the contract checks that the verification key commitment is the same as the program elf.

Reviewer: ```suggestion
Note that the contract checks that the verification key commitment is the same as the program ELF.
```

> ```solidity

Reviewer: ```suggestion

```solidity
```

require(elfCommitment == provingSystemAuxDataCommitment, "ELF does not match");
```

This contract also includes a static call to the Aligned ServiceManager contract 

> to check that the proof was verified in Aligned.

Reviewer: ```suggestion
to check if the proof was verified in Aligned.
```

```solidity
(bool callWasSuccessfull, bytes memory proofIsIncluded) = alignedServiceManager.staticcall(
    abi.encodeWithSignature(
        "verifyBatchInclusion(bytes32,bytes32,bytes32,bytes20,bytes32,bytes,uint256)",
        proofCommitment,
        pubInputCommitment,
        provingSystemAuxDataCommitment,
        proofGeneratorAddr,
        batchMerkleRoot,
        merkleProof,
        verificationDataBatchIndex
    )
);

require(callWasSuccessfull, "static_call failed");

bool proofIsIncludedBool = abi.decode(proofIsIncluded, (bool));
require(proofIsIncludedBool, "proof not included in batch");
```

### Step 3 - Have your app generate the proof and submit it to Aligned

First, generate the proof. 

> For SP1 this means having the [script](../../examples/zkquiz/quiz/script/src/main.rs)

Reviewer: ```suggestion
For SP1, this means having the [script](../../examples/zkquiz/quiz/script/src/main.rs)
```

generate the proof.

Then, submit the proof to Aligned for verification. 
This can be done either with the SDK or by using the Aligned CLI.

> You can find examples of how to submit a proof using the cli 

Reviewer: ```suggestion
You can find examples of how to submit a proof using the CLI 
```

in the [submitting proofs guide](0_submitting_proofs.md).

The call ZK Quiz uses is:

> ```bash

Reviewer: ```suggestion

```bash
```

aligned submit \
    --proving_system SP1 \
    --proof quiz/script/proof-with-io.json \
    --vm_program quiz/program/elf/riscv32im-succinct-zkvm-elf \
    --proof_generator_addr <user_address> \
    --conn wss://batcher.alignedlayer.com
```

### Step 4 - Verify the proof was verified in Aligned

Once the proof is verified in Aligned, 
you can verify that it was verified from your smart contract.

The full example of this flow can be found on the [ZKQuiz Verifier Contract](../../examples/zkquiz/contracts/src/VerifierContract.sol).

> An example [python script](../../examples/zkquiz/encode_verification_data.py) can be found to encode the call data from the json output of the Aligned cli. 

Reviewer: ```suggestion
An example [python script](../../examples/zkquiz/encode_verification_data.py) can be found to encode the call data from the json output of the Aligned CLI.
```

> An example [python script](../../examples/zkquiz/encode_verification_data.py) can be found to encode the call data from the json output of the Aligned cli. 

Reviewer: ```suggestion
An example [python script](../../examples/zkquiz/encode_verification_data.py) can be found to encode the call data from the JSON output of the Aligned CLI. 
```

# Integrating Aligned into your Application

Aligned can be integrated into your applications in a few simple steps to provide a way to verify ZK proofs generated inside your system.

You can find an example of the full flow of using Aligned in your app in the [ZKQuiz example](../../examples/zkquiz).

This example shows a sample app that generates an SP1 proof that a user knows the answers to a quiz and then submits the proof to Aligned for verification. Finally, it includes a smart contract that verifies that a proof was verified in Aligned and mints an NFT.

## 1. Generate your ZK Proof

To submit proofs to Aligned and get them verified, first you need to generate those proofs. Every proving system has its own way of generating proofs.

> You can find examples of how to generate proofs in the [generating proofs guide](3_generating_proofs.md).

Reviewer: ```suggestion
You can find examples on how to generate proofs in the [generating proofs guide](3_generating_proofs.md).
```

> Also you can find an example of the ZKQuiz proof [program](../../examples/zkquiz/quiz/program/src/main.rs) as well as the [script](../../examples/zkquiz/quiz/script/src/main.rs) that generates it in the [ZKQuiz example](../../examples/zkquiz) directory.

Reviewer: ```suggestion
Also you can find an example of the ZKQuiz proof [program](../../examples/zkquiz/quiz/program/src/main.rs) as well as the [script](../../examples/zkquiz/quiz/script/src/main.rs) that generates it in the [ZKQuiz example directory](../../examples/zkquiz).
```

## 2. Write your smart contract

To check if a proof was verified in Aligned, you need to make a call to the `AlignedServiceManager` contract inside your smart contract.

Also, you will need a way to check that the proven program is the correct one.

The Aligned CLI provides a way for you to get the verification key commitment without actually generating and submitting a proof.

You can do this by running the following command:

```bash
aligned get-vk-commitment --input <path_to_input_file>
```

> The following is an example of how to call the `verifyBatchInclusionMethod` from the Aligned ServiceManager contract in your smart contract.

Reviewer: ```suggestion
The following is an example of how to call the `verifyBatchInclusionMethod` from the `AlignedServiceManager` contract in your smart contract.
```

        bytes memory merkleProof,
        uint256 verificationDataBatchIndex
    ) {
        // ... Your function code
        
        require(elfCommitment == provingSystemAuxDataCommitment, "ELF does not match");
        
        (bool callWasSuccessful, bytes memory proofIsIncluded) = alignedServiceManager.staticcall(
            abi.encodeWithSignature(
                "verifyBatchInclusion(bytes32,bytes32,bytes32,bytes20,bytes32,bytes,uint256)",
                proofCommitment,
                pubInputCommitment,
                provingSystemAuxDataCommitment,
                proofGeneratorAddr,
                batchMerkleRoot,
                merkleProof,
                verificationDataBatchIndex
            )
        );

        require(callWasSuccessful, "static_call failed");
        
        bool proofIsIncludedBool = abi.decode(proofIsIncluded, (bool));
        require(proofIsIncludedBool, "proof not included in batch");
        
        // Your function code ...
    }
}
```

You can find an example of the smart contract that checks if the proof was verified in Aligned in the [Quiz Verifier Contract](../../examples/zkquiz/contracts/src/VerifierContract.sol).

> Note that the contract checks that the verification key commitment is the same as the program ELF commitment.

Reviewer: ```suggestion
Note that the contract checks if the verification key commitment is the same as the program ELF commitment.
```

```solidity
require(elfCommitment == provingSystemAuxDataCommitment, "ELF does not match");
```

> This contract also includes a static call to the Aligned ServiceManager contract to check if the proof was verified in Aligned.

Reviewer: ```suggestion
This contract also includes a static call to the `AlignedServiceManager` contract to check if the proof was verified in Aligned.
```

use ethers::prelude::*;
use tokio::time::{sleep, Duration};

async fn wait_for_proof_verification(
    aligned_verification_data: AlignedVerificationData,
    rpc_url: String,
) -> Result<(), anyhow::Error> {
    for _ in 0..10 {
        if verify_proof_onchain(aligned_verification_data.clone(), Chain::Holesky, rpc_url.as_str()).await.is_ok_and(|r| r) {
            println!("Proof verified successfully.");
            return Ok(());
        }
        println!("Proof not verified yet. Waiting 10 seconds before checking again...");
        sleep(Duration::from_secs(10)).await;
    }
    anyhow::bail!("Proof verification failed")
}

#[tokio::main]
async fn main() {
    let aligned_verification_data = // Obtain aligned verification data
    let rpc_url = "https://ethereum-holesky-rpc.publicnode.com".to_string();

    match wait_for_proof_verification(aligned_verification_data, rpc_url).await {
        Ok(_) => println!("Proof verified"),
        Err(err) => println!("Error: {:?}", err),
    }
}
```

You can find an example of the proof submission and verification in the [ZKQuiz Program](../../examples/zkquiz/quiz/script/src/main.rs).

> The example generates a proof, instantiates a wallet to submit the proof, and then submits the proof to Aligned for verification. It then waits for the proof to be verified in Aligned.

Reviewer: ```suggestion
This example generates a proof, instantiates a wallet to submit the proof, and then submits the proof to Aligned for verification. Lastly, it waits for the proof to be verified in Aligned.
```

# Integrating Aligned into your Application

Aligned can be integrated into your applications in a few simple steps to provide a way to verify ZK proofs generated inside your system.

You can find an example of the full flow of using Aligned in your app in the [ZKQuiz example](../../examples/zkquiz).

> This example shows a sample app that generates an SP1 proof that a user knows the answers to a quiz, then submits the proof to Aligned for verification. Finally, it includes a smart contract that verifies that a proof was verified in Aligned and mints an NFT.

Reviewer 2: ```suggestion
This example shows a sample app that generates an SP1 proof that a user knows the answers to a quiz and then submits the proof to Aligned for verification. Finally, it includes a smart contract that verifies that a proof was verified in Aligned and mints an NFT.
```

Generate your ZK proofs using any of the proving systems supported by Aligned.
For this example, we use the SP1 proving system. The current SP1 version used in Aligned is v1.0.8-testnet.

```rust
use sp1_sdk::{ProverClient, SP1Stdin};
use std::io;

const ELF: &[u8] = include_bytes!("../../program/elf/riscv32im-succinct-zkvm-elf");

fn generate_sp1_proof(user_answers: &str) -> Result<Vec<u8>, &'static str> {
    let mut stdin = SP1Stdin::new();
    stdin.write(user_answers);

    let client = ProverClient::new();
    let (pk, vk) = client.setup(ELF);

    match client.prove_compressed(&pk, stdin) {
        Ok(proof) => {
            client.verify_compressed(&proof, &vk).expect("verification failed");
            println!("Proof generated and verified successfully.");
            Ok(proof)
        }
        Err(_) => {
            println!("Proof generation failed. Incorrect answer");
            Err("Proof generation failed")
        }
    }
}

fn main() {
    // Example user answers
    let user_answers = "abc";
    match generate_sp1_proof(user_answers) {
        Ok(proof) => println!("Proof: {:?}", proof),
        Err(err) => println!("Error: {}", err),
    }
}
```

> You can find an example of the quiz proof [program](../../examples/zkquiz/quiz/program/src/main.rs) as well as the [script](../../examples/zkquiz/quiz/script/src/main.rs) that generates it in the [ZKQuiz example](../../examples/zkquiz) directory.

Reviewer 2: This section is a little confusing for someone with no context. The rust code seems to be like a shorter version of the program to generate the proof in `../../examples/zkquiz/quiz/script/src/main.rs`. If we add this code we should add some more context (like why are we showing this code in the first place?) and exaplanation of what is happening

Author: I agree. I will add an explanation on every code section

## 2. Write your smart contract

> To check if a proof was verified in Aligned, you need to call to the Aligned ServiceManager contract inside your smart contract.

Reviewer 2: ```suggestion
To check if a proof was verified in Aligned, you need to make a call to the `AlignedServiceManager` contract inside your smart contract.
```

> Also, you will need a way to check that the proven program is your own.

Reviewer 2: ```suggestion
Also, you will need a way to check that the proven program is the correct one.
```

        bytes memory merkleProof,
        uint256 verificationDataBatchIndex
    ) {
        // ... Your function code
        
        require(elfCommitment == provingSystemAuxDataCommitment, "ELF does not match");
        
        (bool callWasSuccessful, bytes memory proofIsIncluded) = alignedServiceManager.staticcall(
            abi.encodeWithSignature(
                "verifyBatchInclusion(bytes32,bytes32,bytes32,bytes20,bytes32,bytes,uint256)",
                proofCommitment,
                pubInputCommitment,
                provingSystemAuxDataCommitment,
                proofGeneratorAddr,
                batchMerkleRoot,
                merkleProof,
                verificationDataBatchIndex
            )
        );

        require(callWasSuccessful, "static_call failed");
        
        bool proofIsIncludedBool = abi.decode(proofIsIncluded, (bool));
        require(proofIsIncludedBool, "proof not included in batch");
        
        // Your function code ...
    }
}
```

You can find an example of the smart contract that checks if the proof was verified in Aligned in the [Quiz Verifier Contract](../../examples/zkquiz/contracts/src/VerifierContract.sol).

> Note that the contract checks that the verification key commitment is the same as the program ELF.

Reviewer 2: ```suggestion
Note that the contract checks that the verification key commitment is the same as the program ELF commitment.
```

    let client = ProverClient::new();
    let (pk, vk) = client.setup(ELF);

    match client.prove_compressed(&pk, stdin) {
        Ok(proof) => {
            client.verify_compressed(&proof, &vk).expect("verification failed");
            println!("Proof generated and verified successfully.");
            Ok(proof)
        }
        Err(_) => {
            println!("Proof generation failed. Incorrect answer");
            Err("Proof generation failed")
        }
    }
}

fn main() {
    // Example user answers
    let user_answers = "abc";
    match generate_sp1_proof(user_answers) {
        Ok(proof) => println!("Proof: {:?}", proof),
        Err(err) => println!("Error: {}", err),
    }
}
```

You can find an example of the quiz proof [program](../../examples/zkquiz/quiz/program/src/main.rs) as well as the [script](../../examples/zkquiz/quiz/script/src/main.rs) that generates it in the [ZKQuiz example](../../examples/zkquiz) directory.

### Step 2 - Write your smart contract

To check if a proof was verified in Aligned, you need to call to the Aligned ServiceManager contract inside your smart contract.

Also, you will need a way to check that the proven program is your own.

> The aligned CLI provides a way for you to get the verification key commitment without actually generating and submitting a proof.

Reviewer: ```suggestion
The Aligned CLI provides a way for you to get the verification key commitment without actually generating and submitting a proof.
```

        bytes32 proofCommitment,
        bytes32 pubInputCommitment,
        bytes32 provingSystemAuxDataCommitment,
        bytes20 proofGeneratorAddr,
        bytes32 batchMerkleRoot,
        bytes memory merkleProof,
        uint256 verificationDataBatchIndex
    ) {
        // ... Your function code
        
        require(elfCommitment == provingSystemAuxDataCommitment, "ELF does not match");
        
        (bool callWasSuccessful, bytes memory proofIsIncluded) = alignedServiceManager.staticcall(
            abi.encodeWithSignature(
                "verifyBatchInclusion(bytes32,bytes32,bytes32,bytes20,bytes32,bytes,uint256)",
                proofCommitment,
                pubInputCommitment,
                provingSystemAuxDataCommitment,
                proofGeneratorAddr,
                batchMerkleRoot,
                merkleProof,
                verificationDataBatchIndex
            )
        );

        require(callWasSuccessful, "static_call failed");
        
        bool proofIsIncludedBool = abi.decode(proofIsIncluded, (bool));
        require(proofIsIncludedBool, "proof not included in batch");
        
        // Your function code ...
    }
}
```

> You can find the example of the smart contract that checks the proof was verified in Aligned in the [Quiz Verifier Contract](../../examples/zkquiz/contracts/src/VerifierContract.sol).

Reviewer: ```suggestion
You can find an example of the smart contract that checks if the proof was verified in Aligned in the [Quiz Verifier Contract](../../examples/zkquiz/contracts/src/VerifierContract.sol).
```

> Note that the contract checks that the verification key commitment is the same as the program elf.

Reviewer: ```suggestion
Note that the contract checks that the verification key commitment is the same as the program ELF.
```

```solidity
require(elfCommitment == provingSystemAuxDataCommitment, "ELF does not match");
```

This contract also includes a static call to the Aligned ServiceManager contract to check if the proof was verified in Aligned.

```solidity
(bool callWasSuccessfull, bytes memory proofIsIncluded) = alignedServiceManager.staticcall(
    abi.encodeWithSignature(
        "verifyBatchInclusion(bytes32,bytes32,bytes32,bytes20,bytes32,bytes,uint256)",
        proofCommitment,
        pubInputCommitment,
        provingSystemAuxDataCommitment,
        proofGeneratorAddr,
        batchMerkleRoot,
        merkleProof,
        verificationDataBatchIndex
    )
);

require(callWasSuccessfull, "static_call failed");

bool proofIsIncludedBool = abi.decode(proofIsIncluded, (bool));
require(proofIsIncludedBool, "proof not included in batch");
```

### Step 3 - Submit and verify the proof to Aligned

The proof submission and verification can be done either with the SDK or by using the Aligned CLI.

#### Using the SDK

> To submit a proof using the SDK, you can use the `submit` function, and then you can use the `verify_proof_onchain` to check if the proof was correctly verified in Aligned.

Reviewer: ```suggestion
To submit a proof using the SDK, you can use the `submit` function, and then you can use the `verify_proof_onchain` function to check if the proof was correctly verified in Aligned.
```

use ethers::prelude::*;
use tokio::time::{sleep, Duration};

async fn wait_for_proof_verification(
    aligned_verification_data: AlignedVerificationData,
    rpc_url: String,
) -> Result<(), anyhow::Error> {
    for _ in 0..10 {
        if verify_proof_onchain(aligned_verification_data.clone(), Chain::Holesky, rpc_url.as_str()).await.is_ok_and(|r| r) {
            println!("Proof verified successfully.");
            return Ok(());
        }
        println!("Proof not verified yet. Waiting 10 seconds before checking again...");
        sleep(Duration::from_secs(10)).await;
    }
    anyhow::bail!("Proof verification failed")
}

#[tokio::main]
async fn main() {
    let aligned_verification_data = // Obtain aligned verification data
    let rpc_url = "https://ethereum-holesky-rpc.publicnode.com".to_string();

    match wait_for_proof_verification(aligned_verification_data, rpc_url).await {
        Ok(_) => println!("Proof verified"),
        Err(err) => println!("Error: {:?}", err),
    }
}
```

You can find an example of the proof submission and verification in the [Quiz Program](../../examples/zkquiz/quiz/script/src/main.rs).

> The example generates a proof, instantiate a wallet to submit the proof, and then submits the proof to Aligned for verification. It then waits for the proof to be verified in Aligned.

Reviewer: ```suggestion
The example generates a proof, instantiates a wallet to submit the proof, and then submits the proof to Aligned for verification. It then waits for the proof to be verified in Aligned.
```
